docs/topics/js/js-ir-migration.md

```diff
@@ -0,0 +1,186 @@
+templates to automatically mark interfaces as `external`:
+* [Template for `RState`](https://gist.github.com/SebastianAigner/62119536f24597e630acfdbd14001b98)
+* [Template for `RProps`](https://gist.github.com/SebastianAigner/a47a77f5e519fc74185c077ba12624f9)
+
+## Convert properties of external interfaces to var
+
+**Issue**: properties of external interfaces in Kotlin/JS code can't be read-only (`val`) properties because their values can be
+assigned only after the object is created with `jsObject()` or `js()`.
+
+```kotlin
+val myState = js("{}") as CustomComponentState
+myState.name = "name"
+```
+
+**Solution**: convert all properties of external interfaces to `var`:
+
+```kotlin
+// Replace this
+external interface CustomComponentState : RState {
+    val name: String
+}
+```
+
+```kotlin
+// With this
+external interface CustomComponentState : RState {
+    var name: String
+}
+```
+
+## Make boolean properties nullable in external interface
+
+**Issue**: in JavaScript, the `null` or undefined value of a boolean variable equals to `false`. So, boolean properties can be used
+in expressions without being defined. 
+
+```kotlin
+external interface ComponentProps: RProps {
+    var isInitialized: Boolean
+    var visible: Boolean
+}
+```
+
+```kotlin
+val props = js("{}") as ComponentProps
+props.isInitialized = true
+// visible is not initialized - OK in JS – means it's false
+if (props.isInitialized && props.visible) {} // not OK in Kotlin
+```
+
+However, this code with throw a `ClassCastException` because `visible` is `null` while it's type doesn't allow this.
+
+**Solution**: make all `Boolean` properties of external interfaces nullable (`Boolean?`):
+
+```kotlin
+// Replace this
+external interface ComponentProps: RProps {
+    var visible: Boolean
+}
+```
+
+```kotlin
+// With this
+external interface ComponentProps: RProps {
+    var visible: Boolean?
+}
+```
+
+## Convert functions with receiver in external interfaces to regular functions 
+
+**Issue**: external declarations can't contain functions with receiver, such as extension functions or properties with corresponding
+functional types.
+
+**Solution**: convert such functions and properties to regular functions by adding the receiver object as an argument:
+
+```kotlin
+// Replace this
+external interface ButtonProps : RProps {
+    var inside: StyledDOMBuilder<BUTTON>.() -> Unit
+}
+```
+
+```kotlin
+external interface ButtonProps : RProps {
+    var inside: (StyledDOMBuilder<BUTTON>) -> Unit
+}
+```
+
+## Create plain JS objects for interoperability
+
+**Issue**: Properties of a Kotlin object that implements an external interface are not _enumerable_. This means that they are not 
+visible for operations that iterate over the object's properties, such as:
+* `for (var name in obj)`
+* `console.log(obj)`
+* `JSON.stringify(obj)`
+* And other
```

Comment:
I'd get rid of the "and others", and replace "such as" with "for example" before the enumeration